# Financial Consumer Complaints :   Scope of the Project
Consumer complaints on financial products & services for Bank of America from 2017 to 2023, including the dates the complaint was submitted to the CFPB and then sent to the company, the product and issue mentioned in the complaint, and the company's response.

Recommended Analysis

1. Do consumer complaints show any seasonal patterns?

2. Which products present the most complaints? What are its most common issues?

3. How are complaints typically resolved?

4. Can you learn anything from the complaints with untimely responses?


I want to have 3 pages : 
1. Complaints Overview
2. Product & Issue Analysis
3. Geographic & Response Analysis

In [1]:
import pandas as pd 

complaints = pd.read_excel("Consumer_Complaints.xlsx")
complaints.head()

Complaint ID Submitted via Date submitted Date received State  \
0       4848023      Referral     2021-10-24    2021-10-27    NY   
1       3621464           Web     2020-04-24    2020-04-24    FL   
2       5818349           Web     2022-07-27    2022-07-27    CA   
3       7233015      Referral     2023-07-10    2023-07-11    CA   
4       5820224      Referral     2022-07-27    2022-07-28    VA   

                                             Product  \
0                                           Mortgage   
1  Money transfer, virtual currency, or money ser...   
2  Credit reporting, credit repair services, or o...   
3                        Credit card or prepaid card   
4                        Credit card or prepaid card   

                                  Sub-product  \
0                  Conventional home mortgage   
1                   Refund anticipation check   
2                            Credit reporting   
3                General-purpose prepaid card   
4  General-purpose credit card or charge card   

                                               Issue  \
0  Applying for a mortgage or refinancing an exis...   
1                               Lost or stolen check   
2               Incorrect information on your report   
3       Problem getting a card or closing an account   
4                               Closing your account   

                                           Sub-issue  \
0                                                NaN   
1                                                NaN   
2                      Account information incorrect   
3  Trouble getting, activating, or registering a ...   
4                        Company closed your account   

                             Company public response  \
0  Company has responded to the consumer and the ...   
1  Company has responded to the consumer and the ...   
2  Company has responded to the consumer and the ...   
3                                                NaN   
4  Company has responded to the consumer and the ...   

  Company response to consumer Timely response?  
0      Closed with explanation              Yes  
1  Closed with monetary relief              Yes  
2      Closed with explanation              Yes  
3                  In progress              NaN  
4      Closed with explanation              Yes

In [2]:
complaints.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62516 entries, 0 to 62515
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   Complaint ID                  62516 non-null  int64         
 1   Submitted via                 62516 non-null  object        
 2   Date submitted                62516 non-null  datetime64[ns]
 3   Date received                 62516 non-null  datetime64[ns]
 4   State                         62516 non-null  object        
 5   Product                       62516 non-null  object        
 6   Sub-product                   62509 non-null  object        
 7   Issue                         62516 non-null  object        
 8   Sub-issue                     51658 non-null  object        
 9   Company public response       60341 non-null  object        
 10  Company response to consumer  62516 non-null  object        
 11  Timely response?            

In [3]:
complaints.describe(include="all").round()

Complaint ID Submitted via                 Date submitted  \
count        62516.0         62516                          62516   
unique           NaN             7                            NaN   
top              NaN           Web                            NaN   
freq             NaN         45423                            NaN   
mean       4512642.0           NaN  2020-11-24 16:07:14.883869696   
min        2471340.0           NaN            2017-05-01 00:00:00   
25%        3254020.0           NaN            2019-05-22 00:00:00   
50%        4178582.0           NaN            2021-03-02 00:00:00   
75%        5771284.0           NaN            2022-07-14 00:00:00   
max        7458912.0           NaN            2023-08-28 00:00:00   
std        1442917.0           NaN                            NaN   

                        Date received  State                      Product  \
count                           62516  62516                        62516   
unique                            NaN     51                            9   
top                               NaN     CA  Checking or savings account   
freq                              NaN  13709                        24814   
mean    2020-11-25 21:31:05.071341568    NaN                          NaN   
min               2017-05-01 00:00:00    NaN                          NaN   
25%               2019-05-28 00:00:00    NaN                          NaN   
50%               2021-03-03 00:00:00    NaN                          NaN   
75%               2022-07-15 00:00:00    NaN                          NaN   
max               2023-08-28 00:00:00    NaN                          NaN   
std                               NaN    NaN                          NaN   

             Sub-product                Issue                 Sub-issue  \
count              62509                62516                     51658   
unique                46                   76                       158   
top     Checking account  Managing an account  Deposits and withdrawals   
freq               20768                15109                      5596   
mean                 NaN                  NaN                       NaN   
min                  NaN                  NaN                       NaN   
25%                  NaN                  NaN                       NaN   
50%                  NaN                  NaN                       NaN   
75%                  NaN                  NaN                       NaN   
max                  NaN                  NaN                       NaN   
std                  NaN                  NaN                       NaN   

                                  Company public response  \
count                                               60341   
unique                                                  6   
top     Company has responded to the consumer and the ...   
freq                                                60311   
mean                                                  NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   
std                                                   NaN   

       Company response to consumer Timely response?  
count                         62516            61022  
unique                            5                2  
top         Closed with explanation              Yes  
freq                          41044            58619  
mean                            NaN              NaN  
min                             NaN              NaN  
25%                             NaN              NaN  
50%                             NaN              NaN  
75%                             NaN              NaN  
max                             NaN              NaN  
std            

# Plotly Dash Application : Financial Customer Complaint Dashboard

# *************************************************************************************************************************

In [4]:
from dash import Dash, html, dcc, dash_table
from dash.dependencies import Output, Input
from dash.exceptions import PreventUpdate

import dash_bootstrap_components as dbc
from dash_bootstrap_templates import load_figure_template

import plotly.express as px 
import pandas as pd 

# Data Gathering and Cleaning
complaints = pd.read_excel("Consumer_Complaints.xlsx")

# Regarding to the data dictionary, I carry out missing data, to not leave them as there are

complaints["Sub-product"] =  complaints["Sub-product"].fillna("No Sub-product") 
# not all Products have Sub-products

complaints["Sub-issue"] =  complaints["Sub-issue"].fillna("No Sub-issue") 
#not all Issues have corresponding Sub-issues)

complaints["Company public response"] =  complaints["Company public response"].fillna("No response selected yet") 
# The company's optional, public-facing response to a consumer's complaint.

complaints["Timely response?"] = complaints["Timely response?"].fillna("Pending") 
# The status of the response has not yet been determined because the company hasn't provide reponse yet.

dbc_css = "https://cdn.jsdelivr.net/gh/AnnMarieW/dash-bootstrap-templates/dbc.min.css"

app = Dash(__name__, 
           external_stylesheets=[dbc.themes.DARKLY, dbc_css],
           title= "Financial Consumer Complaints",
           suppress_callback_exceptions=True,  # this suppresses the ID not found in layout error
           meta_tags= [{"name": "viewport", 
               "content": "width=device-width, initial-scale=1"}]  
                    # note that meta_tags make your app responsive (web and mobile)
          )

load_figure_template("DARKLY") # Apply the same style to my graphs (figures)


# the style arguments for the sidebar. We use position:fixed and a fixed width
SIDEBAR_STYLE = {
    "position": "fixed",
    "top": 0,
    "left": 0,
    "bottom": 0,
    "width": "20rem",
    "padding": "2rem 1rem",
    "background-color": "#f56942",
}

# the styles for the main content position it to the right of the sidebar and
# add some padding.
CONTENT_STYLE = {
    "margin-left": "18rem",
    "margin-right": "2rem",
    "padding": "2rem 1rem",
}

sidebar = html.Div(
    [
        html.H2([
            html.Span("CFPB", style={"color": "#37de5c"}), 
            html.Br(),
            html.Span("Customer Financial Protection Bureau", style={"fontSize":15})
            
        ], className="display-4"),
        html.Hr(),
        html.P(
            "Consumer complaints on financial products & services for Bank of America from 2017 to 2023 analysis", className="lead"
        ),
        dbc.Nav(
            [
                dbc.NavLink("Overview", href="/", active="exact"),
                dbc.NavLink("Product & Issue", href="/product-issue", active="exact"),
                dbc.NavLink("Geographic Distribution", href="/geographic", active="exact"),
                
            ],
            vertical=True,
            pills=True,
        ),
    ],
    style=SIDEBAR_STYLE,
)


content = html.Div(id="page-content", style=CONTENT_STYLE)

app.layout = html.Div([dcc.Location(id="url"), sidebar, content])



# Static figures :


# Overview Static figures
# Static figures should be placed outside the callback function, they do need update based on user inputs

# Complaints by Submitted via (submission channels)
complaints_channels = (complaints
                       .groupby("Submitted via")
                       .agg({"Complaint ID": "count"})
                       .reset_index()
                       .rename(columns={"Complaint ID": "Total Complaints"})
                       .sort_values("Total Complaints", ascending=False)
                      )

complaints_channels["Percentage"] =  ((complaints_channels["Total Complaints"]
                                      /complaints_channels["Total Complaints"].sum() * 100
                                      ).round(2).astype(str) + '%')



fig_complaints_channels = px.funnel(complaints_channels, 
          y="Submitted via", 
          x="Total Complaints",
          hover_data= {
              "Submitted via" : True,
              "Total Complaints" : True,
              "Percentage" : ':.2f' + '%'  # Format percentage to 2 decimal places
          },
#           color_discrete_sequence= px.colors.qualitative.Prism,
          title= "Total Complaints Received by Channel Type"
        
            
         ).update_layout(title={ 
                                "font":{"size":20, "family":"Arial"}
                               })


# Complaints Response Distribution By Reponse Status

response_status = complaints["Timely response?"].value_counts().reset_index()
fig_response_status = px.pie(response_status, 
       names="Timely response?", 
       values="count",
       hole=.5,
       category_orders= {"Timely response?": ["Yes", "No", "Pending"]},
       title= "Complaints Responses Distribution By Response Status"
                        
               
      )



# ********************************************************************************************************************
# Static figure, product and issue page

complaints_product =  complaints.Product.value_counts().rename("Total Complaints").reset_index()
fig_complaints_product = px.bar(complaints_product.sort_values(by="Total Complaints"), 
       x="Total Complaints", 
       y="Product", 
       title="Total Complaints By Product"
      )

# issue 
complain_issue = (complaints.Issue
                  .value_counts()
                  .rename("Total Complaints")
                  .reset_index()
                  .sort_values(by="Total Complaints", ascending=False)
                 )
fig_complain_issue = px.bar(complain_issue.iloc[:5].sort_values(by="Total Complaints"), 
       x="Total Complaints", 
       y="Issue", 
       title="Total Complaints By Issue Type"
      )

# Sub-products
complaints_sub_produts = (complaints[(complaints["Sub-product"] != "No Sub-product")]["Sub-product"]
                          .value_counts()
                          .rename("Total Complaints")
                          .reset_index()
                          .sort_values(by="Total Complaints", ascending=False)
                          .iloc[:10]
                          .sort_values("Total Complaints")
                         )
fig_complaints_sub_produts = px.bar(complaints_sub_produts, 
                                    x="Total Complaints",
                                    y="Sub-product", 
                                    title="Total Complaints by Sub-products - Top 10"
                                   )


# ****************************************************************************************************************
# Geographic Distribution figure -static figure

complaints_state = complaints["State"].value_counts().rename("Total Complaints").reset_index()
fig_complaints_state = px.choropleth(
    complaints_state,
    locations="State",
    locationmode="USA-states",
    color= "Total Complaints",
    scope= "usa",
    hover_name= "State",
    custom_data= ["State"],
    title= "Total Complaints By State",
             )


top5_state = (complaints_state
              .sort_values(by="Total Complaints", ascending=False)
              .iloc[:5]
              .sort_values(by="Total Complaints")
             )
fig_top5_state = px.bar(top5_state, 
       x="Total Complaints", 
       y="State",
       title= "Top 5 States By Complaints"
      )


# CFPB sent the complaint to the company
complaints["Processing Time (hour)"] = ((complaints["Date received"] - complaints["Date submitted"])
                                        .dt.total_seconds() / 3600)


avg_processing_time_state = (complaints[complaints["Timely response?"].isin(["Yes", "No"])]
                       .groupby("State")["Processing Time (hour)"]
                       .mean()
                       .reset_index()
                       .rename(columns={"Processing Time (hour)": "Average Processing Time (hour)"})
                      )


fig_avg_processing_time_state = px.bar(avg_processing_time_state, 
       x="State", 
       y="Average Processing Time (hour)",
       title= "CFPB sent complaint the company : Average Processing Time (hour) By State",
       hover_data= {
           "State" : True,
           "Average Processing Time (hour)" : ":.2f"
       }
      )


timely_response_state = (complaints
                         .loc[complaints["Timely response?"] == "Yes"]
                         .groupby("State", as_index=False)
                         .agg({"Timely response?":"count"})
                         .sort_values(by="Timely response?", ascending=False)
                        )

fig_timely_response_state = px.bar(timely_response_state, x="State", 
                                   y="Timely response?", 
                                   title="Timely Response By State")


response_consumer = complaints["Company response to consumer"].value_counts().reset_index()
fig_response_consumer = px.bar(response_consumer, x="Company response to consumer", 
       y="count", 
       title="Count of Company response Types to consumer",
      )


# KPI # KPI Cards Visuals  : total_complaints, number_of_state, number_of_product, number_of_subproduct,
# timely_response_rate, average_processing_time_

total_complaints = complaints["Complaint ID"].count() 

timely_response_rate = ((complaints["Timely response?"] == "Yes").sum() / total_complaints * 100).round(2)

average_processing_time_ = (complaints["Processing Time (hour)"].mean()/24).round(2)
number_of_state = complaints["State"].nunique()

# Number of Product 
number_of_product = complaints.Product.nunique()
number_of_subproduct = complaints[(complaints["Sub-product"] != "No Sub-product")]["Sub-product"].nunique()




# ***********************************************************************************************************************
# Pages contents


# Overview page

overview_page_layout = dbc.Container(
#     style = {"margin-left": "25px"},
    children = [
        

    dbc.Row([
        dbc.Col([
            dbc.Card([
                dbc.CardHeader("Total Complaints"),
                dbc.CardBody(
                    [
                        html.H5(f"{total_complaints:,.0f}".replace(",", " "), className="card-title", style={"text-align":"center"}),
                    ]
                ),

            ])
        ], width=2),
        
        dbc.Col([
            dbc.Card([
                dbc.CardHeader("Number of State"),
                dbc.CardBody(
                    [
                        html.H5(f"{number_of_state}", className="card-title", style={"text-align":"center"}),
                    ]
                ),

            ])
        ], width=2),
        
        dbc.Col([
            dbc.Card([
                dbc.CardHeader("Total Products"),
                dbc.CardBody(
                    [
                        html.H5(f"{number_of_product}", 
                                className="card-title", style={"text-align":"center"}),
                       
                    ]
                ),

            ])
        ], width=2),
        
         dbc.Col([
            dbc.Card([
                dbc.CardHeader("Total Sub-Products"),
                dbc.CardBody(
                    [
                        html.H5(f"{number_of_subproduct}", 
                                className="card-title", style={"text-align":"center"}),
                       
                    ]
                ),

            ])
        ], width=2),
        
         dbc.Col([
            dbc.Card([
                dbc.CardHeader("Average Processing Time"),
                dbc.CardBody(
                    [
                        html.H5(f"{average_processing_time_} days", className="card-title", style={"text-align":"center"}),
                    ]
                ),

            ])
        ], width=2),
        
        
         dbc.Col([
            dbc.Card([
                dbc.CardHeader("Timely Response Rate"),
                dbc.CardBody(
                    [
                        html.H5(f"{timely_response_rate} %", className="card-title", style={"text-align":"center"}),
                    ]
                ),

            ])
        ], width=2),
        
        
    ]),
    
    html.Br(),
    
    dbc.Row([
        dbc.Col([
            dbc.Card(
                [
                    dcc.Dropdown(id="time-component", 
                                 options={
                                     "month" : "Month Level",
                                     "year" : "Year Level",
                                     "date" : "Day Level"
                                 },
                                 value="month",
                                 className = "dbc"
                                ),
                    dcc.Graph(id="Complaints-Trend")
                ], 
                body=True)
        ]),
        
          dbc.Col([
            dbc.Card(
                [
                    dcc.Dropdown(id="time-component-2", 
                                 options={
                                     "month" : "Month Level",
                                     "year" : "Year Level"
                                 },
                                 value="month",
                                 className = "dbc"
                                ),
                    dcc.Graph(id="Change-Trend")
                ], 
                body=True)
        ]),
        
    ]),
    
    html.Br(),
        
    dbc.Row([
        dbc.Col([
            dbc.Card(dcc.Graph(figure=fig_complaints_channels), body=True)
        ]),
        
        dbc.Col([
            dbc.Card(dcc.Graph(figure=fig_response_status), body=True)
        ]),
    ])
        
        
        
        
])




# Product & Issue Analysis


product_issue_page_layout = dbc.Container(
#     style = {"margin-left": "25px"},
    children = [
    
    dbc.Row([
        
        dbc.Col([
            dbc.Card(dcc.Graph(figure=fig_complaints_product), body=True,)
        ]),
    
    ]),
        
    html.Br(),
    dbc.Row([
        dbc.Col([
            dbc.Card(dcc.Graph(figure=fig_complaints_sub_produts), body=True,)
        ]),
    ]),
        
    
    html.Br(),
    
        
    dbc.Row([
        dbc.Col([
            dbc.Card(dcc.Graph(figure=fig_complain_issue), body=True)
        ]),
    ])
              
        
])


# Geographic & Response Analysis


geographic_response_page_layout = dbc.Container(
#     style = {"margin-left": "25px"},
    children = [
        
        dbc.Row([
            dbc.Col([
                dbc.Card(dcc.Graph(id="cross-filter-Complaints-State-Map", 
                                   figure=fig_complaints_state,
                                   hoverData = {"points" : [{"customdata": ["CA"]}]}
                                  ), body=True)
            ]),

             dbc.Col([
                dbc.Card(dcc.Graph(figure=fig_top5_state), body=True)
            ]),
        ]),

        
        html.Br(),

        dbc.Row([
            dbc.Col([
                dbc.Card([
                    dcc.Dropdown(id="time-component-state", 
                                 options={
                                     "month" : "Month Level",
                                     "year" : "Year Level",
                                 },
                                 value="month",
                                 className = "dbc"
                                ),
                    dcc.Graph(id="complaints-trend-state"),
                ], body=True)
            ]),
        ]),
        
         html.Br(),

        dbc.Row([
            dbc.Col([
                dbc.Card(dcc.Graph(id="processing-time", figure=fig_avg_processing_time_state), body=True)
            ]),
        ]),
        
        html.Br(),
        
        dbc.Row([
            dbc.Col([
                dbc.Card(dcc.Graph(figure=fig_timely_response_state), body=True)
            ]),

             dbc.Col([
                dbc.Card(dcc.Graph(figure=fig_response_consumer), body=True)
            ]),
        ]),
              
        
])






# Sidebar callback, to render pages

@app.callback(
    Output(component_id="page-content", component_property="children"), 
    [Input(component_id="url", component_property="pathname")]
)
def render_page_content(pathname):
    if pathname == "/":
        return overview_page_layout
    elif pathname == "/product-issue":
        return product_issue_page_layout
    elif pathname == "/geographic":
        return geographic_response_page_layout
    # If the user tries to reach a different page, return a 404 message
    return html.Div(
        [
            html.H1("404: Not found", className="text-danger"),
            html.Hr(),
            html.P(f"The pathname {pathname} was not recognised..."),
        ],
        className="p-3 bg-light rounded-3",
    )



# Overview page callback function



@app.callback(
    Output(component_id="Complaints-Trend", component_property="figure"),
    Input(component_id="time-component", component_property="value")
)
def show_trend_overview(selection):
    if not selection:
        raise PreventUpdate
        
    # Define a mapping for grouping functions
    grouping_map = {
        "date": complaints["Date received"].dt.date,
        "month": complaints["Date received"].dt.month,
        "year": complaints["Date received"].dt.year
    }
    
    # Group by month/year/day number and count complaints
    trend_complaints = (
        complaints
        .groupby(grouping_map[selection]) # complaints["Date received"].dt.month/year
        .agg({"Complaint ID": "count"})
        .reset_index()
        .rename(columns={"Date received" : f"{selection}", "Complaint ID": "Total Complaints"}))
    
    
    # Rename month numbers to month names if selection is month
    if selection == "month":
        trend_complaints[selection] = trend_complaints[selection].map({ 
            1: "January", 2: "February", 3: "March", 4: "April", 
            5: "May", 6: "June", 7: "July", 8: "August", 
            9: "September", 10: "October", 11: "November", 12: "December"
        })  # my data is sorted by month number that way.

    # current vs previous month complaints : 
    trend_complaints = trend_complaints.assign(
        previous_ = trend_complaints["Total Complaints"].shift(1),
    )
    
    
    
    if selection == "year":
        fig_current_previous = px.bar(
            trend_complaints[trend_complaints[selection] > trend_complaints[selection].min()], 
            x=selection, 
            y=["Total Complaints", "previous_"],
            barmode= "group"
        )
    else :
        fig_current_previous = px.line(
            trend_complaints,
            x=selection,
            y=["Total Complaints", "previous_"],
        )
    
    fig_current_previous = fig_current_previous.update_layout(
        title= f"Current vs Previous {selection.capitalize()} Total Complaints ",
        legend_title = "",
    )
    
    return fig_current_previous


@app.callback(
   
    Output(component_id="Change-Trend", component_property="figure"), 
    Input(component_id="time-component-2", component_property="value")
)
def show_change_overview(selection):
    if not selection:
        raise PreventUpdate
        
        
    # Define a mapping for grouping functions
    grouping_map = {
        "date": complaints["Date received"].dt.date,
        "month": complaints["Date received"].dt.month,
        "year": complaints["Date received"].dt.year
    }
    
    # Group by month/year number and count complaints
    trend_complaints = (
        complaints
        .groupby(grouping_map[selection]) # complaints["Date received"].dt.month/year
        .agg({"Complaint ID": "count"})
        .reset_index()
        .rename(columns={"Date received" : f"{selection}", "Complaint ID": "Total Complaints"}))
    
    
    # Rename month numbers to month names if selection is month
    if selection == "month":
        trend_complaints[selection] = trend_complaints[selection].map({ 
            1: "January", 2: "February", 3: "March", 4: "April", 
            5: "May", 6: "June", 7: "July", 8: "August", 
            9: "September", 10: "October", 11: "November", 12: "December"
        })  # my data is sorted by month number that way.


    # current vs previous month complaints : 
    trend_complaints = trend_complaints.assign(
        change_over_ = (trend_complaints["Total Complaints"].diff(1) / 
               trend_complaints["Total Complaints"].shift(1) * 100).round(2)
    )
    
    
    trend_complaints['Color'] = trend_complaints['change_over_'].apply(lambda x: 'green' if x < 0 else 'red')

    category_order  = trend_complaints[selection].to_list() 
    
    fig_change = px.bar(trend_complaints, 
           x=selection, 
           y="change_over_", 
           color="Color", 
           color_discrete_map={'red': 'red', 'green': 'green'},
           category_orders= {f"{selection}" : category_order}
        ).update_layout(showlegend=False)
    
    fig_change = fig_change.update_layout(
        title = f"Total Complaints: {selection.capitalize()} Over {selection.capitalize()} Change (%)"
    )
    
    return fig_change


# Geographic section callback function
@app.callback(
    Output(component_id="complaints-trend-state", component_property="figure"),
    Input(component_id="cross-filter-Complaints-State-Map", component_property="hoverData"),
    Input(component_id="time-component-state", component_property="value")
)
def plot_complaints_in_state(hoverData, selection):
    state_name = hoverData["points"][0]["customdata"][0] # This grabs the “State” value from the hoverData dictionary
    df_state_selected = complaints.query("State == @state_name")
    
       # Define a mapping for grouping functions
    grouping_map = {
        "month": df_state_selected["Date received"].dt.month,
        "year": df_state_selected["Date received"].dt.year
    }
    
    # Group by month/year and count complaints
    df_state_selected_complaints = (
        df_state_selected
        .groupby(grouping_map[selection])
        .agg({"Complaint ID": "count"})
        .reset_index()
        .rename(columns={"Date received" : f"{selection}", "Complaint ID": "Total Complaints"}))
    
    
    # Rename month numbers to month names if selection is month
    if selection == "month":
        df_state_selected_complaints[selection] = df_state_selected_complaints[selection].map({ 
            1: "January", 2: "February", 3: "March", 4: "April", 
            5: "May", 6: "June", 7: "July", 8: "August", 
            9: "September", 10: "October", 11: "November", 12: "December"
        })  # my data is sorted by month number that way.

   
    
    if selection == "year":
        fig_df_state_selected_complaints = px.bar(
            df_state_selected_complaints, 
            x=selection, 
            y="Total Complaints"
        )
        
    else :
        fig_df_state_selected_complaints  = px.area(
            df_state_selected_complaints, 
            x= selection, 
            y="Total Complaints"
    )
    
    fig_df_state_selected_complaints =  (fig_df_state_selected_complaints.update_layout(
                                             title= {
                                                 "text" : f"Total Complaints by {selection} in {state_name}",
#                                                  "x": 0.5, # center the title
#                                                  "y" : 0.87
                                             },
                                             yaxis_title="Count")
                                        )
    
    return fig_df_state_selected_complaints


        



if __name__ == "__main__":
    app.run_server(debug=True, port=2004)



# Key takeways : 

### My goal for this project is to Analyze complaint volume across time and by product, issue, geographic area, channel Type, see how complaints are resolved, what the company response to each complaints it received from the CFPB.
* CFPB has registred in total 62 516 Consumer complaints on financial products & services for Bank of America from 2017 to 2023.
* The complaints concerns about 9 products and 46 sub-products in 52 US states.
* CFPB takes about 1.22 days to send customer complaint to the relevant company from the day it received it 
* The timely response rate to complaints is 93.77 % 
* From the overall years (2017-2023), we registred more customers complaints in 2022. 
* We noticed a peak on the month of july, that means CFPB received the huge volume of complaints on this month.
* In February we have seen a decrease of complaints around 7%, and since then it shows in increase of customers complaints until june, where it slightly decreased ( around 3%) before we noticed the peak in July. The two following month the volum of complaints shows an interesting decrease. In October, the volume rises slightly, then declines in November and and starts to rise again in December.
* Most of the complaints is submitted via web, representing around 73% of the overall volume of complaints
* On the products side, Checking or Savings Acount received the highest number of complaints (24814) , which could indicate issues in banking operations or customer dissatisfaction with this product. Credit Card and Prepaid  Card also has a high  complaints volume (16197).
* On the Sub-products side, most of complaints are abount checking account (20768 complaints in total, this is around 33% of the overall volume)
* Managing account is the most frequently issue reported.
* The five states with the most complaints are California (CA), Florida (FL), Texas (TX), New York (NY) and Georgia (GA); with California recording the most.
* In California, the higest number of complaints is recorded in 2021. 
* The highest average processing time (CFPB to send complaints to company) is in South Dakota (SD) (74.12 hours - more than 3 days), and the lowest is in Iowa (IA) (7.46 hours - less than one day).
* California receives the most complaints, but also the most timely responses.
* Most of the companies reponses to customers complaints are accompanied with explanation.